external/warc/src/main/java/com/digitalpebble/stormcrawler/warc/WARCSpout.java

```diff
@@ -0,0 +1,530 @@
+                Long.toString(response.date().toEpochMilli()));
+        if (response.ipAddress().isPresent()) {
+            metadata.addValue(
+                    protocolMDprefix + ProtocolResponse.RESPONSE_IP_KEY,
+                    response.ipAddress().get().getHostAddress());
+        }
+        if (request != null) {
+            metadata.addValue(
+                    protocolMDprefix + ProtocolResponse.REQUEST_HEADERS_KEY,
+                    httpHeadersVerbatim(request));
+        }
+        metadata.addValue(
+                protocolMDprefix + ProtocolResponse.RESPONSE_HEADERS_KEY,
+                httpHeadersVerbatim(http));
+    }
+
+    @SuppressWarnings({ "rawtypes", "unchecked" })
+    @Override
+    public void open(Map conf, TopologyContext context,
+            SpoutOutputCollector collector) {
+        _collector = collector;
+        record = Optional.empty();
+
+        maxContentSize = ConfUtils.getInt(conf, "http.content.limit", -1);
+        if (contentBufferSize > maxContentSize) {
+            // no need to buffer more content than max. used
+            contentBufferSize = maxContentSize;
+        }
+        storeHTTPHeaders = ConfUtils.getBoolean(conf, "http.store.headers",
+                false);
+        protocolMDprefix = ConfUtils.getString(conf,
+                ProtocolResponse.PROTOCOL_MD_PREFIX_PARAM, protocolMDprefix);
+        sleepEmitFetched = ConfUtils.getLong(conf,
+                "warc.spout.emit.fetched.sleep.ms", 0);
+
+        int metricsTimeBucketSecs = ConfUtils.getInt(conf,
+                "fetcher.metrics.time.bucket.secs", 10);
+        eventCounter = context.registerMetric("warc_spout_counter",
+                new MultiCountMetric(), metricsTimeBucketSecs);
+    }
+
+    @Override
+    public void nextTuple() {
+        if (!active)
+            return;
+
+        if (buffer.isEmpty()) {
+            try {
+                populateBuffer();
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        }
+
+        if (warcReader == null && buffer.isEmpty()) {
+            // input exhausted
+            try {
+                Thread.sleep(10);
+            } catch (InterruptedException e1) {
+            }
+            return;
+        }
+
+        if (!record.isPresent())
+            nextRecord();
+
+        while (record.isPresent() && !isHttpResponse(record)) {
+            String warcType = record.get().type();
+            if (warcType == null) {
+                LOG.warn("No type for {}", record.get().getClass());
+            } else {
+                eventCounter.scope(
+                        "warc_skipped_record_of_type_" + record.get().type())
+                        .incr();
+                LOG.debug("Skipped WARC record of type {}",
+                        record.get().type());
+            }
+            if (storeHTTPHeaders && "request".equals(warcType)) {
```

Comment:
that string "request" is used in other places in the module - have it as a constant of enum somewhere?

Reply:
Ok. But maybe better use `instanceof WarcRequest` as the literal "request" is an internal of jwarc.